----
#**Comprehensions in Python: Lists, Dictionaries, Sets, and Generators**

1. **List Comprehension:**
   - *Description:* List comprehensions are a concise way to create lists in Python. They provide a compact syntax for generating new lists by applying an expression to each item in an iterable, optionally filtering items based on a condition.
   - *Use Case:* List comprehensions are useful when you want to create a new list by transforming or filtering elements from an existing iterable.

2. **Dictionary Comprehension:**
   - *Description:* Dictionary comprehensions allow you to create dictionaries using a compact syntax. They define key-value pairs by applying expressions to items in an iterable, with an optional condition for filtering.
   - *Use Case:* Dictionary comprehensions are handy when you need to create dictionaries from an iterable, such as mapping values to keys or filtering data.

3. **Set Comprehension:**
   - *Description:* Set comprehensions provide a way to create sets efficiently. Like list comprehensions, they apply an expression to each item in an iterable and allow for filtering based on a condition.
   - *Use Case:* Set comprehensions are useful when you want to create unique sets of values from an iterable while applying transformations or filters.

4. **Generator Comprehension:**
   - *Description:* Generator comprehensions offer a memory-efficient way to create generators, which are iterable objects. They generate values on-the-fly, applying expressions to items in an iterable and optionally filtering elements.
   - *Use Case:* Generator comprehensions are suitable when you want to work with large data sets or create iterators that don't load all values into memory at once.
----


In [52]:
# List Comprehension
for i in range(100):
  print(i, end = ', ')

0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 

In [53]:
# Applying list comprehension to above loop

lst_comp1 = [i for i in range(100)]
print(lst_comp1)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99]


In [54]:
# Print all numbers from 0 to 99, divisible by three

# Regular Method:

for i in range(100):
  if i % 3 == 0:
    print(i, end = ',')

0,3,6,9,12,15,18,21,24,27,30,33,36,39,42,45,48,51,54,57,60,63,66,69,72,75,78,81,84,87,90,93,96,99,

In [55]:
# List Comprehension Method

lst_comp2 = [i for i in range(100) if i % 3 == 0]
print(lst_comp2)
print(type(lst_comp2))

[0, 3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39, 42, 45, 48, 51, 54, 57, 60, 63, 66, 69, 72, 75, 78, 81, 84, 87, 90, 93, 96, 99]
<class 'list'>


In [56]:
# For String datas
# Regular Method:

fruits_lst = ['apple','banana','cherries','dates','eggfruit','fig','graphs']
for i in fruits_lst:
  if'a' in i:
    print(i)

apple
banana
dates
graphs


In [57]:
# List Comprehension Method

fruits_lst = ['apple','banana','cherries','dates','eggfruit','fig','graphs']

lst_comp3 = [i for i in fruits_lst if 'a' in i]
print(lst_comp3)

['apple', 'banana', 'dates', 'graphs']


In [58]:
# In-class Practice Assignment:

fruits_lst = ['apple','banana','cherries','dates','eggfruit','fig','graphs']
# Regular Method
for i in fruits_lst:
  if 'apple' not in i:
    print(i)


banana
cherries
dates
eggfruit
fig
graphs


In [59]:
fruits_lst = ['apple','banana','cherries','dates','eggfruit','fig','graphs']

lst_comp4 = [i for i in fruits_lst if 'apple' not in i]
print(lst_comp4)

['banana', 'cherries', 'dates', 'eggfruit', 'fig', 'graphs']


In [60]:
# All the fruits that ARE NOT of apple kind

apples_fruits_lst = ['apple','custard apple','green apple','dates','eggfruit','fire red apple','graphs']

lst_comp4 = [i for i in fruits_lst if 'apple' not in i]
print(lst_comp4)

['banana', 'cherries', 'dates', 'eggfruit', 'fig', 'graphs']


In [61]:
# All the fruits that ARE of the apple kind

apples_fruits_lst = ['apple','custard apple','green apple','dates','eggfruit','fire red apple','graphs']

lst_comp5 = [i for i in apples_fruits_lst if 'apple' in i]
print(lst_comp5)

['apple', 'custard apple', 'green apple', 'fire red apple']


In [62]:
# All the fruits that ARE of the apple kind
# Using: Print statement inside the list

apples_fruits_lst = ['apple','custard apple','green apple','dates','eggfruit','fire red apple','graphs']

lst_comp5 = [print(i) for i in apples_fruits_lst if 'apple' in i]


apple
custard apple
green apple
fire red apple


In [63]:
# Practice Example
# Regular Method

types = ['indian', 'american', 'nepali']
fruits = ['apple', 'mango', 'pineapple']

for i in types:
  for j in fruits:
    if 'apple' in j:
      print(i, j)

indian apple
indian pineapple
american apple
american pineapple
nepali apple
nepali pineapple


In [64]:
# List Comprehension Method

types = ['indian', 'american', 'nepali']
fruits = ['apple', 'mango', 'pineapple']

lst_comp6 = [[i,j] for i in types for j in fruits if 'apple' in j]
lst_comp6

[['indian', 'apple'],
 ['indian', 'pineapple'],
 ['american', 'apple'],
 ['american', 'pineapple'],
 ['nepali', 'apple'],
 ['nepali', 'pineapple']]

In [65]:
# Practice Question: Print all the items in list starts with 'c' or 'C'
# Regular Method

names = ['Ch', 'Dh','Eh', 'cb', 'Tb', 'Td']

for i in names:
  if i.startswith('C') or i.startswith('c'):
    print(i)



Ch
cb


In [66]:
# List Comprehension Method
names = ['Ch', 'Dh','Eh', 'cb', 'Tb', 'Td']

lst_comp7 = [i for i in names if i.startswith('C') or i.startswith('c')]
print(lst_comp7)
print(type(lst_comp7))


['Ch', 'cb']
<class 'list'>


In [67]:
lst_matrix = [i for i in range(3)]
print(lst_matrix)


[0, 1, 2]


In [68]:
lst_matrix0 = [i for i in range(3) for j in range(3)]
print(lst_matrix0)
print()

lst_matrix = [[i for i in range(3)] for j in range(3)]
print(lst_matrix)

[0, 0, 0, 1, 1, 1, 2, 2, 2]

[[0, 1, 2], [0, 1, 2], [0, 1, 2]]


In [69]:
# Regular Implementation

for j in range(3):
  for i in range(3):
    print(i)


0
1
2
0
1
2
0
1
2


In [70]:
lst_str = [i for i in 'Python Programming']
lst_str

['P',
 'y',
 't',
 'h',
 'o',
 'n',
 ' ',
 'P',
 'r',
 'o',
 'g',
 'r',
 'a',
 'm',
 'm',
 'i',
 'n',
 'g']

In [71]:
# Practice Question: Print all numbers from 0 to 100 that are both divisible by 5 and 10 using nested loop

# List Comprehension Method

lst1 = [i for i in range(101) if i % 5 == 0 if i % 10 == 0]
print(lst1)

[0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100]


In [72]:
# Regular Implementation:

for i in range(101):
  if i % 5 == 0:
    if i % 10 == 0:
      print(i , end = ', ')


0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 

In [73]:
# Using 2D array
ndim2array = [[10, 20, 30], [40, 50, 60], [70, 80, 90]]

# Create a new 2D array with elements that are the product of the corresponding elements from ndim2array
lst1 = [[i * j for i in row] for j in ndim2array for row in ndim2array]

print(lst1)


[[[10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30], [10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30], [10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30]], [[10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 20, 30, 10, 2

In [74]:
# 2D array
ndim2array = [[10, 20, 30], [40, 50, 60], [70, 80, 90]]

# Iterate over rows and columns of the 2D array and print the product of corresponding elements
for row in ndim2array:
    for element in row:
        print(element, end=" ")
    print()

# Print the original 2D array and its type
print(ndim2array)
print(type(ndim2array))


10 20 30 
40 50 60 
70 80 90 
[[10, 20, 30], [40, 50, 60], [70, 80, 90]]
<class 'list'>


**Dictionary Comprehension**

In [75]:
# Regular

for i in range(10):
  print(f"{i}: Item{i}")

print(type(i))


0: Item0
1: Item1
2: Item2
3: Item3
4: Item4
5: Item5
6: Item6
7: Item7
8: Item8
9: Item9
<class 'int'>


In [76]:
dictCompre = {i : f"Item{i}" for i in range(10)}
dictCompre

{0: 'Item0',
 1: 'Item1',
 2: 'Item2',
 3: 'Item3',
 4: 'Item4',
 5: 'Item5',
 6: 'Item6',
 7: 'Item7',
 8: 'Item8',
 9: 'Item9'}

**Set Comprehension**

In [77]:
# Set Comprehension Implementastion

#set_values = {'apple', 'ball', 'pineapple', 'apple', 'banana', 'banana'}
dict_values = ['apple', 'ball', 'pineapple', 'apple', 'banana', 'banana']


unique_stuffs = {i for i in dict_values}
print(unique_stuffs)
print(type(unique_stuffs))

print()
print('-Comparing to Dictionary Comprehension-')
print()

non_unique_stuffs = [i for i in dict_values]
print(non_unique_stuffs)
print(type(non_unique_stuffs))


{'banana', 'ball', 'pineapple', 'apple'}
<class 'set'>

-Comparing to Dictionary Comprehension-

['apple', 'ball', 'pineapple', 'apple', 'banana', 'banana']
<class 'list'>


In [78]:
# Regular Implementation

#set_values = {'apple', 'ball', 'pineapple', 'apple', 'banana', 'banana'}
dict_values = ['apple', 'ball', 'pineapple', 'apple', 'banana', 'banana']

unique_stuffs = set()

for i in dict_values:
  unique_stuffs.add(i)

print(unique_stuffs)
print(type(unique_stuffs))

{'banana', 'ball', 'pineapple', 'apple'}
<class 'set'>


**In-Class Assignment**

In [79]:
# List Comprehensions

# Task: Print length of each word in a sentence

sentence = "This is a data science course in Takeo"

# Regular Implementation:

for i in sentence.split():
  print(len(i) , end = ' ')

# List comprehension Method

length = [len(i) for i in sentence.split()]
print("\nThe length for each word is ", length)
print(type(length))


4 2 1 4 7 6 2 5 
The length for each word is  [4, 2, 1, 4, 7, 6, 2, 5]
<class 'list'>


In [80]:
sentence = "first second third"
word1 = sentence.split()
print(word1)
type(word1)


['first', 'second', 'third']


list

In [81]:
# Dictionary Comprehension
# Task: Split each word in the string sentence and put them in a dictionary by: key: value -> word -> length of word

sentence2 = "This is a data science course in Takeo"

word_length_regular = {}

# Regular Implementation:

for word in sentence2.split():
  word_length_regular[word] = len(word)

print("Regular Method: ",word_length_regular)

# Dictionary Comprehension Method:

word_length_comp = {word:len(word) for word in sentence2.split()}
print("Comprehension Method: ",word_length_comp)
print(type(word_length_comp))


Regular Method:  {'This': 4, 'is': 2, 'a': 1, 'data': 4, 'science': 7, 'course': 6, 'in': 2, 'Takeo': 5}
Comprehension Method:  {'This': 4, 'is': 2, 'a': 1, 'data': 4, 'science': 7, 'course': 6, 'in': 2, 'Takeo': 5}
<class 'dict'>


In [82]:
# Set Comprehension

# Regular implementation:

# Two list:

num_divisible_by5 = [5,10,15,20,25,30]
my_favourite_numbers = [3,4,7,8, 15, 30, 25]

union_set = set()   # Empty set
list_union = []

for num1 in num_divisible_by5:
  union_set.add(num1)
  list_union.append(num1)

for num2 in my_favourite_numbers:
  union_set.add(num2)
  list_union.append(num2)

print("List: ", list_union)


print("Regular Implement:", union_set)

# Set Comprehension Method:

union_set2 = {numa for numa in num_divisible_by5}.union({numb for numb in my_favourite_numbers})

print('Union Method:', union_set2)
print(type(union_set2))


List:  [5, 10, 15, 20, 25, 30, 3, 4, 7, 8, 15, 30, 25]
Regular Implement: {3, 4, 5, 7, 8, 10, 15, 20, 25, 30}
Union Method: {3, 4, 5, 7, 8, 10, 15, 20, 25, 30}
<class 'set'>


In [83]:
# List Comprehension vs Generator Comprehensions

# List
numbers = [i for i in range(100)]
print(numbers)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99]


In [90]:
# Generator: Does not store data in memory, making it memory-efficient
from timeit import timeit

# Creating a generator comprehension for numbers from 0 to 99
gen_comp = (i for i in range(100))
print(type(gen_comp))  # Confirms the type as <class 'generator'>

# Measuring execution time for list comprehension (faster but memory-intensive)
list_comp_time = timeit("[i for i in range(100)]")
print("Time for List Comprehension:", list_comp_time)

# Measuring execution time for generator comprehension (slower but memory-efficient)
gen_comp_time = timeit("[(i for i in range(100))]")
print("Time for Generator Comprehension:", gen_comp_time)


<class 'generator'>
4.693191003999999
0.7018502660000081


The results:
- Time for List Comprehension: 4.693191003999999
- Time for Generator Comprehension: 0.7018502660000081

These results illustrate the performance difference between list comprehensions (faster but memory-intensive) and generator comprehensions (slower but memory-efficient). Generator comprehensions are valuable when working with large datasets that don't fit entirely in memory, optimizing memory usage and processing speed.